### ISSUE TITLE
Plugin API raises DeprecationWarning in python3.x

### ISSUE THREAD

**Issue by PyB1l** [bottlepy/bottle#1070]
When you want to use **context** parameter of the **apply** method, python3  interpreter compains about this: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec().

**Comment by PyB1l** [comment]
Fixing it is rather trivial, I have dealt with the same issue on a bottle Plugin I've written myself. I could create a pull request if you think that is worth resolving this.

**Comment by defnull** [comment]
Sure, as long as the change is backwards compatible with all supported python versions (>=2.7.3, >=3.3)

**Comment by defnull** [comment]
Dropping Python 3.3 support would be okay, though, if that helps.

**Comment by Cologler** [comment]
raise `ValueError` if we added annotations on callback:

``` shell
.venv\lib\site-packages\bottle.py:557: DeprecationWarning: inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
  return getargspec(self.get_undecorated_callback())[0]
*** ValueError: Function has keyword-only parameters or annotations, use inspect.signature() API which can support them
```

**Comment by drougge** [comment]
I made a trivial pull request to fix this. https://github.com/bottlepy/bottle/pull/1381

**Comment by defnull** [comment]
Fixed in 0.12.23 and master (for some time now). Thanks :)
